Refactor result details modal

naegelyd opened the pull request:
This removes the knowledge capture components from the modal and also removes all the HTML construction from the JS in favor of using Marionette views and HTML templates.

naegelyd committed:
Refactor summary and effects details sections

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor phenotypes details section

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor prediction scores details section

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor frequencies details section

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor articles details section

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor Cohorts details section

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Use id over class for result details dialog

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Fix cohort popover CSS to scroll overflow

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Hide overflow in modal window

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Fix code allowing for expandable details rows

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor solvebio/clinvar details section

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/js/src/ui/modals/result.js:
```diff
@@ -3,531 +3,13 @@
+    '../variant'
+], function($, _, Backbone, Marionette, utils, variant) {
 
+    /*
```
Should this commented out block still be here?

bruth commented on varify/static/js/src/ui/modals/result.js:
```diff
@@ -768,84 +79,44 @@ define([
-            _.bindAll(this, 'onSaveError', 'onSaveSuccess');
-
-            this.assessmentTab = new AssessmentTab;
+        selectors: {
```
Is this a new convention you are starting for selectors external to the view? A comment should be added to make note of this.

bruth replied:
Ideally, in the future the coupling should be refactoring even through events or some parent view.

naegelyd replied:
I only did this out of convenience and for clarify. I cannot use the standard `ui` convention because the actual elements are not yet available even though they exist in the `el`. That is, if I tried to put all the `selectors` into the `ui` block like is normal convention, then lines like this in `open` fail because the ui element is empty:

``` javascript
...
ui: {
    rows: '[data-target=expandable-row]'
},
...
open: fucntion() {
    ...
    this.ui.rows.css('height', '' + this.maxExpandableHeight + 'px')
        .css('overflow', 'hidden');
    ...
```

I'm not trying to start a new trend or anything. I was just trying to group all the selectors I was using into a single property for clarity. If there is a better approach, just let me know and I will update to use it instead. Otherwise, I will just add a comment to explain the appraoch to the code itself.

naegelyd committed:
Refactor the pathogenicity/categort assessment metrics details

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Refactor assessment metrics table in details modal

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add selectors explanation and move local selector to ui

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Remove unused template methods

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Fix bad indents in variant resources file

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/js/src/ui/modals/result.js:
```diff
@@ -896,67 +116,94 @@ define([
+            }));
+
+            this.frequencies.show(new variant.Frequencies({
+                model: new Backbone.Model(this.model.get('variant'))
```
Is there any value in initializing a single model for the `variant`? This is being used a couple times above. The `Variant` model could define all the sub-{models, collections} for the nested structures such cohorts, etc. This makes it a bit more clear what embedded data is supported and enables implementing sub-document parsing on nested models/collections. For example:

``` javascript
var Variant = Backbone.Model.extend({
    constructor: function(attrs, options) {
        // Ensure to parse..
        options = _.defaults({parse: true}, options);

        this.cohorts = new Backbone.Collection();
        this.frequencies = new Backbone.Collection();
        // etc..

        Backbone.Model.prototype.constructor.call(this, attrs, options);
    },
    parse: function(attrs) {
        this.frequencies.reset(attrs.frequencies);
        this.cohorts.reset(attrs.cohorts);
        // etc..
        return attrs;
    }
});
```

bruth replied:
Each of this sub-{models, collections} would correspond to the views they are being passed into.

bruth commented on varify/static/js/src/ui/modals/result.js:
```diff
@@ -896,67 +116,94 @@ define([
             // Reset the row and item heights and overflow styles as they may
             // have been toggled previously.
-            $('[data-target=expandable-details-row]').css('height', '' + this.maxExpandableHeight + 'px')
+            this.$el.find(this.ui.expandableRows)
```
This looks to be redundant since `this.ui.expandableRows` is already a jQuery object and is scoped to `this.$el`.

bruth commented on varify/static/js/src/ui/modals/result.js:
```diff
@@ -896,67 +116,94 @@ define([
+                .css('height', '' + this.maxExpandableHeight + 'px')
                 .css('overflow', 'hidden');
-            $('[data-target=expand-collapse-link]').text(this.showMoreText);
+            this.$el.find(this.selectors.expandCollapseLink)
```
Should this be _found_ in this element? Isn't the point of the `this.selectors` hash to be external to this element?

bruth commented on varify/static/js/src/ui/tables/row/genes.js:
```diff
@@ -0,0 +1,29 @@
+    });
+
+    var GeneList = Marionette.CollectionView.extend({
+        template: function() {},
```
`CollectionView`s do not need a `template` defined.

naegelyd committed:
Remove selectors hash in favor of ui and rebinding

Since many of the ui elements we need to address are not available in
the templates, they will not be bound during the normal pre-render
bindUIElements call. As a result, we rebind the UI elements after all
the regions are shown or in the onRender method before interacting with
the ui object.

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Remove empty templates from CollectionViews

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Use Variant model to simplify result modal setup

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/js/src/models/variant.js:
```diff
@@ -0,0 +1,52 @@
+            if (attrs.solvebio && attrs.solvebio.clinvar) {
+                clinvarResults = attrs.solvebio.clinvar.results;
+            }
+            clinvarResults = clinvarResults || [];
```
Small nit, but `reset` does not require an array to be passed.

bruth commented on varify/static/js/src/ui/modals/result.js:
```diff
@@ -896,67 +113,85 @@ define([
+            // be found during the normal UI binding call. We rebind here so
+            // since all the elements should now be available as all the
+            // regions have been populated and their views shown.
+            this.bindUIElements();
```
Do any of the selectors depends on collection view items? If so, could there be a race condition with the items rendering before these elements are rebound?

naegelyd replied:
No. None of the selectors depend on collection view items. They are all in the templates of the different regions themselves so there should be no race condition here.

bruth commented on varify/static/templates/variant/clinvar-item.html:
```diff
@@ -0,0 +1,12 @@
+
+<ul>
+    <li>
+        <small>Siginificance</small> <b><%= data.assertion.clinicalsignificance %></b>
```
Significance.

bruth commented on varify/static/templates/variant/clinvar-item.html:
```diff
@@ -0,0 +1,12 @@
+Assertion: <a target='_blank' href='https://www.ncbi.nlm.nih.gov/clinvar/'<%= data.assertion.rcvaccession %>/'><%= data.assertion.rcvaccession %></a>
```
I am getting a complaint in the console that `data.assertion` is undefined.

naegelyd committed:
Remove unneccessary assignment before clinvarResults reset

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Fix spelling error in clinvar template

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Fix bad references in clinvar item template

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Detach popover click handlers in onClose

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/js/src/ui/variant/assessment-metrics.js:
```diff
@@ -0,0 +1,256 @@
+        initialize: function() {
+            _.bindAll(this, 'hidePopover');
+
+            $(document).on('click', this.hidePopover);
```
This should go in `onRender` since the handler depends on UI elements to be rendered.

bruth commented on varify/static/js/src/ui/variant/assessment-metrics.js:
```diff
@@ -0,0 +1,256 @@
+        initialize: function() {
+            _.bindAll(this, 'hidePopover');
+
+            $(document).on('click', this.hidePopover);
```
Same as above, put in `onRender`.

bruth commented on varify/static/js/src/ui/variant/assessment-metrics.js:
```diff
@@ -0,0 +1,256 @@
+        },
+
+        onRender: function() {
+            this.bindUIElements();
```
What is the purpose of doing this explicitly?

naegelyd replied:
The popovers are not part of this view's template but are instead on the itemviews. Because of that, the initial UI binding will not find the popovers. So, we need to bind the UI elements after the itemviews have rendered so that the popovers will be found.

bruth commented on varify/static/js/src/ui/variant/assessment-metrics.js:
```diff
@@ -0,0 +1,256 @@
+        },
+
+        onRender: function() {
+            this.bindUIElements();
```
Same as above, why do this explicitly?

bruth commented on varify/static/js/src/ui/variant/assessment-metrics.js:
```diff
@@ -0,0 +1,256 @@
+                throw new Error('Variant ID Required');
+            }
+
+            this.model= new models.AssessmentMetrics({}, {
```
Space before `=`. `null` can be passed to the model rather than initializing an object.

bruth commented on varify/static/js/src/ui/variant/cohorts.js:
```diff
@@ -0,0 +1,107 @@
+        initialize: function() {
+            _.bindAll(this, 'hidePopover');
+
+            $(document).on('click', this.hidePopover);
```
Put in `onRender`.

bruth commented on varify/static/js/src/ui/variant/cohorts.js:
```diff
@@ -0,0 +1,107 @@
+        },
+
+        onRender: function() {
+            this.bindUIElements();
```
Remove assuming it's redundant from the above questions.

naegelyd committed:
Move click handler registration into onRender

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/js/src/ui/variant/effects.js:
```diff
@@ -0,0 +1,138 @@
+                gene = transcript.gene;
+            }
+
+            var proteinChange, changeString =
```
Reading this was a tad deceiving, consider moving `changeString` to next line with it's assignment.

naegelyd committed:
Fix formatting of assessment metrics initialization

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Clarify assignments of variables used in effects region

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/js/src/ui/variant/phenotypes.js:
```diff
@@ -0,0 +1,89 @@
+        tagName: 'li',
+
+        serializeData: function() {
+            var data = Marionette.ItemView.prototype.serializeData.apply(
```
All this does it call `toJSON()` on the model or collection that is bound. You can leave it if you want.
